src/Serialise/Cborg/Tutorial.hs

```diff
@@ -16,33 +16,239 @@ module Serialise.Cborg.Tutorial
+of reasons,
+
+ * Deviating from the type-guided encoding that the 'Generic' instance will
+   provide
+
+ * Interfacing with other CBOR implementations
+
+ * Enabling migrations for future changes to the type or its encoding
+
+A minimal hand-rolled instance will define the 'encode' and 'decode' methods,
+
+> instance Serialise Animal where
+>     encode = encodeAnimal
+>     decode = decodeAnimal
+
+Below we will describe how to write these pieces.
 
 -}
 
 {- $encoding
 
-Lorem ipsum...
+For the purposes of encoding, abstract CBOR resentations are represented by the
+'Serialise.Cborg.Encoding.Tokens' type. Such a representation can be efficiently
+built using the 'Serialise.Cborg.Encoding.Encoding' 'Monoid'.
+
+For instance, to implement an encoder for the @Animal@ type above we might write,
+
+> encodeAnimal :: Animal -> Encoding
+> encodeAnimal (HoppingAnimal name height) =
+>     encodeListLen 3 <> encodeTag 0 <> encode name <> encode height
+> encodeAnimal (WalkingAnimal name speed) =
+>     encodeListLen 3 <> encodeTag 1 <> encode name <> encode speed
+
+Here we see that each encoding begins with a /length/, describing how many
+values belonging to our @Animal@ will follow. We then encode a /tag/, which
+identifies which constructor. We then encode the fields.
 
 -}
 
 {- $decoding
 
-Lorem ipsum...
+Decoding CBOR representations to Haskell values is done in the 'Decoder'
+'Monad'. We can write a 'Decoder' for the @Animal@ type defined above as
+follows,
+
+> decodeAnimal :: Decoder s Animal
+> decodeAnimal = do
+>     len <- decodeListLen
+>     tag <- decodeTag
+>     case (len, tag) of
+>       (3, 0) -> HoppingAnimal <$> decode <*> decode
+>       (3, 1) -> WalkingAnimal <$> decode <*> decode
+>       _      -> fail "invalid Animal encoding"
+
+-}
+
+{- $migrations
+
+One eventuality that data serialisation schemes need to account for is the need
+for changes in the data's structure. There are two types of compatibility
+which we might want to strive for in our serialisers,
+
+ * Backward compatibility, such that newer versions of the serialiser can read
+   older versions of an encoding
+
+ * Forward compatibility, such that older versions of the serialiser can read
+   (or at least tolerate) newer versions of an encoding
+
+Below we will look at a few of the types of
+changes which we may need to make and describe how these can be handled in a
+backwards-compatible manner with @cborg@.
+
+*** Adding a constructor
+
+Say we want to add a new constructor to our @Animal@ type, @SwimmingAnimal@,
+
+> data Animal = HoppingAnimal { animalName :: String, hoppingHeight :: Int }
+>             | WalkingAnimal { animalName :: String, walkingSpeed :: Int }
+>             | SwimmingAnimal { numberOfFins :: Int }
+>             deriving (Generic)
+
+We can account for this in our hand-rolled serialiser by simply adding a new tag
+to our encoder and decoder,
+
+> encodeAnimal :: Animal -> Encoding
+> {- ... other cases are unchanged -}
```

Comment:
Hmm, alright. I was thinking that including them would make it harder to spot the differences, but I suppose I'll try adding them and see how it looks.